Framework/Algorithms/src/SANSSolidAngle.cpp

```diff
@@ -0,0 +1,184 @@
+  V3D inPlane = V3D(sampleDetVec);
+  inPlane.setY(0.0);
+
+  // This is the angle between the sample-to-detector vector
+  // and its project on the X-Z plane.
+  return sampleDetVec.angle(inPlane);
+}
+
+/**
+ *Returns correct angular function for the specified type
+ */
+std::function<double(int64_t)>
+getAngularFunction(const SpectrumInfo &spectrumInfo, const std::string &type) {
+  if (type == "Normal") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosTheta = std::cos(spectrumInfo.twoTheta(histogramIndex));
+      return cosTheta * cosTheta * cosTheta;
+    };
+  } else if (type == "Tube") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosTheta = std::cos(spectrumInfo.twoTheta(histogramIndex));
+      double cosAlpha = std::cos(getYTubeAngle(spectrumInfo, histogramIndex));
+      return cosTheta * cosTheta * cosAlpha;
+    };
+  } else if (type == "Wing") {
+    return [&spectrumInfo](int64_t histogramIndex) {
+      double cosAlpha = std::cos(getYTubeAngle(spectrumInfo, histogramIndex));
+      return cosAlpha * cosAlpha * cosAlpha;
+    };
+  } else {
+    throw std::runtime_error("Invalid type of correction");
+  }
+}
+
+/**
+ * Compute the solid angle using the correct angular function
+ */
+class CalculateSolidAngle {
+public:
+  CalculateSolidAngle(const double pixelArea, const SpectrumInfo &spectrumInfo,
+                      const ComponentInfo &componentInfo,
+                      std::function<double(int64_t)> &&f)
+      : m_pixelArea(pixelArea), m_spectrumInfo(spectrumInfo),
+        m_componentInfo(componentInfo), m_f(std::move(f)) {}
+  double operator()(int64_t histogramIndex) const {
+    Kernel::V3D scaleFactor = m_componentInfo.scaleFactor(histogramIndex);
+    double scaledPixelArea = m_pixelArea * scaleFactor[0] * scaleFactor[1];
+    double angularTerm = m_f(histogramIndex);
+    double distance = m_spectrumInfo.l2(histogramIndex) * 1.e3; // mm
+    return (scaledPixelArea * angularTerm) / (distance * distance);
+  }
+
+private:
+  const double m_pixelArea;
+  const SpectrumInfo &m_spectrumInfo;
+  const ComponentInfo &m_componentInfo;
+  const std::function<double(int64_t)> m_f;
+};
+
+void SANSSolidAngle::exec() {
+
+  // Get the workspaces
+  API::MatrixWorkspace_const_sptr inputWS = getProperty("InputWorkspace");
+  const int64_t numberOfSpectra =
+      static_cast<int64_t>(inputWS->getNumberHistograms());
+  API::MatrixWorkspace_sptr outputWS =
+      WorkspaceFactory::Instance().create(inputWS, numberOfSpectra, 2, 1);
+  // The result of this will be a distribution
+  outputWS->setDistribution(true);
+  outputWS->setYUnit("");
+  outputWS->setYUnitLabel("Steradian");
+  setProperty("OutputWorkspace", outputWS);
+
+  const auto &spectrumInfo = inputWS->spectrumInfo();
+  const auto &componentInfo = inputWS->componentInfo();
+  const double pixelSizeX =
+      inputWS->getInstrument()->getNumberParameter("x-pixel-size")[0];
```

Comment:
I think this is the problem for generic data. There must be a way to get the pixel size from the instrument geometry itself, rather than parameters, can you investigate?

Reply:
Moreover, one of our instruments, for example, has different banks, where the pixel sizes are different. Hence, the sizes have to be retrieved from the geometry.

Reply:
Instrument 2.0 doesn't seem to provide methods for this information. I *think* this is the source of your segfault (the parameter probably doesn't exit for your instrument).

Reply:
The segfault is probably due to missing parameter. But for a regular SANS instrument the pixel sizes can be retrieved from the bounding box.

Reply:
That should be the fall back place so users can override it if they think they know better.